src/newtondynamics_physics/SysNewton.cpp

```diff
@@ -0,0 +1,428 @@
+
+using osp::active::SysPhysics;
+
+using osp::active::ACompHierarchy;
+using osp::active::ACompRigidBody;
+using osp::active::ACompShape;
+using osp::active::ACompSolidCollider;
+using osp::active::ACompTransform;
+using osp::active::ACompTransformControlled;
+using osp::active::ACompTransformMutable;
+
+using osp::Matrix4;
+using osp::Vector3;
+
+// Callback called for every Rigid Body (even static ones) on NewtonUpdate
+void cb_force_torque(const NewtonBody* pBody, dFloat timestep, int threadIndex)
+{
+    // Get Scene from Newton World
+    ActiveScene* scene = static_cast<ActiveScene*>(
+                NewtonWorldGetUserData(NewtonBodyGetWorld(pBody)));
+
+    // Get Associated entity
+    uintptr_t userData = reinterpret_cast<uintptr_t>(NewtonBodyGetUserData(pBody));
+    ActiveEnt ent = static_cast<ActiveEnt>(userData);
+
+    auto &rBody = scene->reg_get<ACompRigidBody>(ent);
+    auto &rTransformComp = scene->reg_get<ACompTransform>(ent);
+
+    // Check if transform has been set externally
+    auto &tfMutable = scene->reg_get<ACompTransformMutable>(ent);
+    if (tfMutable.m_dirty)
+    {
+        // Set matrix
+        NewtonBodySetMatrix(pBody, rTransformComp.m_transform.data());
+
+        tfMutable.m_dirty = false;
+    }
+
+    // TODO: deal with changing inertia, mass or stuff
+
+    // Apply force and torque
+    NewtonBodySetForce(pBody, rBody.m_netForce.data());
+    NewtonBodySetTorque(pBody, rBody.m_netTorque.data());
+
+    // Reset accumolated net force and torque for next frame
+    rBody.m_netForce = {0.0f, 0.0f, 0.0f};
+    rBody.m_netTorque = {0.0f, 0.0f, 0.0f};
+}
+
+void SysNewton::add_functions(ActiveScene &rScene)
+{
+    SPDLOG_LOGGER_INFO(rScene.get_application().get_logger(),
+                      "Initing Sysnewton");    rScene.debug_update_add(rScene.get_update_order(), "physics", "wire", "",
+                            &SysNewton::update_world);    //NewtonWorldSetUserData(m_nwtWorld, this);
+
+    // Connect signal handlers to destruct Newton objects when components are
+    // deleted.
+
+    rScene.get_registry().on_destroy<ACompNwtBody>()
+                    .connect<&SysNewton::on_body_destruct>();
+
+    rScene.get_registry().on_destroy<ACompNwtCollider>()
+                    .connect<&SysNewton::on_shape_destruct>();
+
+    rScene.get_registry().on_destroy<ACompNwtWorld>()
+                    .connect<&SysNewton::on_world_destruct>();
+}
+
+void SysNewton::update_world(ActiveScene& rScene)
+{
+    ActiveReg_t &rReg = rScene.get_registry();
+
+    // Try getting the newton world from the scene root.
+    // It is very unlikely that multiple physics worlds are ever needed
+    auto *worldComp = rReg.try_get<ACompNwtWorld>(rScene.hier_get_root());
+
+    if (worldComp == nullptr)
+    {
+        return; // No physics world component
+    }
+
+    if (worldComp->m_nwtWorld == nullptr)
+    {
+        // Create world if not yet initialized
+        worldComp->m_nwtWorld = NewtonCreate();
```

Comment:
Why wouldn't the object creating the ACompNwtWorld be responsible for this?

Reply:
Hmm, good thought. The code creating the scene also has to specify adding Newton systems, so this isn't much of a problem. Though I guess it would be cleaner if it followed the same scheme as the other Newton components.

Reply:
in terms of performance, i would wager that having it be pre-created, as opposed to checking if it's valid and creating it new if it's not, would be faster. One less branch.